### PR TITLE
fix: update reference to Pug version for Set up a Template Engine challenge

### PR DESCRIPTION
**Pull request by jeremyfiel**
Relates to https://github.com/freeCodeCamp/boilerplate-advancednode/pull/9.

Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #XXXXX

This is to update the Pug dependency version in the instructions for this challenge. This PR is a dependent of the PR referenced in the advanced boilerplate repo. I completely updated the boilerplate including the dependencies.

Thanks!

### CODE REVIEW

**Comment by scissorsneedfoodtoo** [comment]
This change looks good and partially addresses https://github.com/freeCodeCamp/freeCodeCamp/issues/38404. 

I believe we'll need to modify a current test or add a new test to fully address the problem where `res.sendFile(...)` passes even though the instructions say to change that method to `res.render(...)`. 

The problem seems to be that the test suite doesn't recognize the difference between `sendFile` and `render`, and probably receives the same data either way. We'll need to add or modify a test so it's something like the second `testString` here that does a regex check on what's in `server.js`: 

```js
testString: getUserInput => $.get(getUserInput('url')+ '/_api/server.js') .then(data => { assert.match(data, /('|")view engine('|"),( |)('|")pug('|")/gi, 'Your project should set Pug as a view engine'); }, xhr => { throw new Error(xhr.statusText); })
```

We'll need to check that people use `res.render(...)` directly.

But that can be handled in another PR. Would you be interested in working on that? Everything here LGTM.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Could you remove the changes to `package-lock.json`?

LGTM, otherwise.

**Comment by jeremyfiel** [comment]
> Could you remove the changes to `package-lock.json`?
> 
> LGTM, otherwise.

@ojeytonwilliams  i tried updating and pushing another commit but it seems like it didn't take it.  I'm thinking the build is modifying it some how.  What do you think?

**Comment by ojeytonwilliams** [comment]
@jeremyfiel I'm not quite sure what was going wrong, but this is how I fixed it:

```sh
git rebase master
git checkout master package-lock.json
git commit --amend
```

Possibly it had been reintroduced during the merge.  Not sure!

**Comment by jeremyfiel** [comment]
> This change looks good and partially addresses #38404.
> 
> I believe we'll need to modify a current test or add a new test to fully address the problem where `res.sendFile(...)` passes even though the instructions say to change that method to `res.render(...)`.
> 
> The problem seems to be that the test suite doesn't recognize the difference between `sendFile` and `render`, and probably receives the same data either way. We'll need to add or modify a test so it's something like the second `testString` here that does a regex check on what's in `server.js`:
> 
> ```js
> testString: getUserInput => $.get(getUserInput('url')+ '/_api/server.js') .then(data => { assert.match(data, /('|")view engine('|"),( |)('|")pug('|")/gi, 'Your project should set Pug as a view engine'); }, xhr => { throw new Error(xhr.statusText); })
> ```
> 
> We'll need to check that people use `res.render(...)` directly.
> 
> But that can be handled in another PR. Would you be interested in working on that? Everything here LGTM.

i'm not entirely sure we need to assert using render if the boilerplate is updated to use it. 
The instructions state the following line but i've already updated the boilerplate to use render and the rest of the exercises never use sendFile anyways.  This seems like we should update the boilerplate to set the method to render but maybe leave the path parameter blank and get the user to update it to "index.pug" to point to the correct files to render. 

Another option is we can leave the block inside the get method blank and ask them to update the entire block.  
either of these makes more sense the way the test is written and we can always update the tests to further assert the use of render if the code block is left blank.

Something I just noticed is the path is actually completely misleading because it's using .html instead of .pug.  I can see how that would be confusing. 

> Lastly, you should change your response to the request for the index route to res.render with the path to the view views/pug/index.pug.
>app.route('/')
  .get((req, res) => {
    res.sendFile(process.cwd() + '/views/index.html');  *****Should we remove this sample code***?** 
  });

My proposal is: 
>app.route('/')
  .get((req, res) => {
  // Write your code here to correctly render the index page
  });

then modify the tests to assert render is used and the path is matching the expectation.

**Comment by ojeytonwilliams** [comment]
> My proposal is:
> 
> > app.route('/')
> > .get((req, res) => {
> > // Write your code here to correctly render the index page
> > });
> 
> then modify the tests to assert render is used and the path is matching the expectation.

Sure, though I think it's important that the page does not hang.  How about:

```js
app.route('/')
.get((req, res) => {
   res.send('replace this!'); // Write your code here to correctly render the index pag
});
```

or similar.  Just something to make sure that their app works straight away.

**Comment by jeremyfiel** [comment]
> > My proposal is:
> > > app.route('/')
> > > .get((req, res) => {
> > > // Write your code here to correctly render the index page
> > > });
> > 
> > 
> > then modify the tests to assert render is used and the path is matching the expectation.
> 
> Sure, though I think it's important that the page does not hang. How about:
> 
> ```js
> app.route('/')
> .get((req, res) => {
>    res.send('replace this!'); // Write your code here to correctly render the index pag
> });
> ```
> 
> or similar. Just something to make sure that their app works straight away.

i tested it on the original glitch repo and it shows "Site didn't respond" which could confuse campers although it replies with 200, I doubt most are logging the response codes to understand that aspect.  It's not finding the root page because the route is empty. 
If we set it to res.sendFile with an invalid parameter they are actually getting an error instead.  I think we can explain this a bit in the article.  Something to the affect,

>Your page will not load until you correctly _render_ the index file. Using the "Show" menu at the top of the Glitch page will allow you to view your app next to your code. This is helpful to see if your page is loading as expected. If you are seeing the page with a message "Site didn't respond", this indicates your routing is incomplete on the root path. You can use a relative path, absolute path or native Node methods such as ["process.cwd()"](https://nodejs.org/docs/latest-v12.x/api/process.html) to point the route to the index file.  

I think asking a camper to hunt for native node methods is a bit far fetched vs asking them to review the express docs for the render function, although this is supposed to be "Advanced Node".  hopefully the hint of saying "render" will be enough and I can link to the Node methods.

Because this is the boilerplate for the beginning of Advanced Node, the app is currently not in a working state anyways. This setup work is the starting point.
I've read a lot of complaints over the past year where campers are very confused with the advanced node coursework. I hope this can be a new beginning to really set them up for success. It will require going through the other boiler plates to align them.

**Comment by ojeytonwilliams** [comment]
> I think asking a camper to hunt for native node methods is a bit far fetched vs asking them to review the express docs for the render function, although this is supposed to be "Advanced Node". hopefully the hint of saying "render" will be enough and I can link to the Node methods.

That's a very good point.  I just tried this out and had to go to the express docs because `render('views/pug/index.pug')` does not work and the instructions imply that you should use `views/pug/index.pug`.  My preference would be to guide people to the concise solution `render('pug/index')`, rather than using any node methods.  Perhaps the instructions could be:

> Lastly, you should change your response to the request for the index route to `render` the file in `views/pug/index.pug`.  `res.render()` takes a relative path (relative to `views`) as its argument and does not require an extension.

We could link to the express docs, too.  Generally we don't use external links, but when we're relying on a library I think it's fine.

**Comment by jeremyfiel** [comment]
> > I think asking a camper to hunt for native node methods is a bit far fetched vs asking them to review the express docs for the render function, although this is supposed to be "Advanced Node". hopefully the hint of saying "render" will be enough and I can link to the Node methods.
> 
> That's a very good point. I just tried this out and had to go to the express docs because `render('views/pug/index.pug')` does not work and the instructions imply that you should use `views/pug/index.pug`. My preference would be to guide people to the concise solution `render('pug/index')`, rather than using any node methods. Perhaps the instructions could be:
> 
> > Lastly, you should change your response to the request for the index route to `render` the file in `views/pug/index.pug`.  `res.render()` takes a relative path (relative to `views`) as its argument and does not require an extension.
> 
so actually the relative path would only work if the server and index were in the same directory, sub directories don't count.  that's not the case so the absolute path can be achieved a few different ways. The main point is we need to drill down into directories because we are starting at the base path /app. we can confirm this by logging process.cwd() // /app.  Using the extension doesn't really matter although the express docs state you don't need it.  So I guess the tests need to be updated to allow for either some of these, or all of these variations. This is a pretty heavy regex request.  I'm not sure if there is a better way to test if the page rendered successfully vs actually parsing the server.js for the matching data.  I found an example on stack using mocha/chai where we can just test if the page loaded and read some text on the page from the index.pug. This seems much simpler than expecting a bunch of text variations in server.js.
```js
describe('Home Page', () => {
    it('This is the Home page', (done)=> {
        chai.request(server)
        .get('/home')
        .end((error, respond) => {            
            expect(/Welcome/, done);                     
        done();
        });
    }); 
})
```
```
//All of these work
/app/views/pug/index
/app/views/pug/index.pug
../views/pug/index
../views/pug/index.pug
./pug/index
./pug/index.pug
${process.cwd()}/views/pug/index
${process.cwd()}/views/pug/index.pug

```
> We could link to the express docs, too. Generally we don't use external links, but when we're relying on a library I think it's fine.

Ok.

**Comment by jeremyfiel** [comment]
I added a new test for {issues/38404} and reworded the instructions a bit to make more sense and lead the camper to a successful result without giving them the answer.

**Comment by scissorsneedfoodtoo** [comment]
@jeremyfiel and @ojeytonwilliams, awesome discussion. I'm learning a lot, so thank you for all the details you've included here.

@jeremyfiel, like you mentioned, it does make sense to change the boilerplate here to more effectively lead learners to the right solution, especially since it's the first challenge in this section. And I think you're right, we shouldn't have to rely on regex testing if we set things up so learners have to use `.render()` from the beginning. It'll make things much easier and cleaner from a testing perspective, and we can use a test like the one you posted above that checks for specific test on the home page.

Given all the possible ways to render the page, maybe the text could be something like:

"Lastly, you should change your response to the request for the index route to render the file in `views/pug/index.pug`. `res.render()` takes a string of a file path an argument. The path can be a relative path (relative to `views`), or an absolute path, and does not require an extension."

**Comment by jeremyfiel** [comment]
> @jeremyfiel and @ojeytonwilliams, awesome discussion. I'm learning a lot, so thank you for all the details you've included here.
> 
> @jeremyfiel, like you mentioned, it does make sense to change the boilerplate here to more effectively lead learners to the right solution, especially since it's the first challenge in this section. And I think you're right, we shouldn't have to rely on regex testing if we set things up so learners have to use `.render()` from the beginning. It'll make things much easier and cleaner from a testing perspective, and we can use a test like the one you posted above that checks for specific test on the home page.
> 
> Given all the possible ways to render the page, maybe the text could be something like:
> 
> "Lastly, you should change your response to the request for the index route to render the file in `views/pug/index.pug`. `res.render()` takes a string of a file path an argument. The path can be a relative path (relative to `views`), or an absolute path, and does not require an extension."

so I actually added the test using regex to verify the server.js has the expected input.  There is already another test checking for data returned on the template after it loads. I went the regex direction because we are trying to coerce campers into specific methods being used in the actual code, rather than only checking the result. If we were in a production environment, this type of test wouldn't be necessary.
Thanks for the comments! Appreciate it!

**Comment by jeremyfiel** [comment]
@ojeytonwilliams  i updated the instructions here too so they coincide with the boilerplate.

File: curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md [hunk 1]
```diff
@@ -6,21 +6,28 @@ forumTopicId: 301564
...
+To set up <em>Pug</em> for use in your project, you will need to add it as a dependency in your package.json. Don't forget to add the name of the package and the version. Use the package.json for some examples of the correct syntax.
+
+Express needs to know which template engine you are using. We will use the <b>set</b> method to assign 'pug' as the <b>'view-engine'</b>. <code>app.set('view engine', 'pug')</code>
+
+Your page will not load until you correctly <em>render</em> the index file in our <em>'views/pug'</em> directory. Using the "Show" menu at the top of the Glitch page will allow you to view your app next to your code. This is helpful to see if your page is loading as expected. If you are seeing the page with a message "Pug template is not defined.", this indicates your routing is incomplete on the root path ('/'). You can get a <em>response</em> using a relative path to point the root ('/') to the index file.
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md:17]:
```suggestion
Your page will not load until you correctly <em>render</em> the index file in our <em>'views/pug'</em> directory. Using the "Show" menu at the top of the Glitch page will allow you to view your app next to your code. This is helpful to see if your page is loading as expected. If you are seeing the page with a message "Pug template is not defined.", you need to change your response to the request for the <code>/</code> route to render the file in <code>views/pug/index.pug</code>. <code>res.render()</code> takes a string of a file path an argument. The path can be a relative path (relative to views), or an absolute path, and does not require an extension. 
```
I liked @scissorsneedfoodtoo's wording, so I incorporated it here.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Nearly there! I think we need to go into the details of how to to `render` a file, to avoid people giving up at the first challenge.

File: curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md (hunk 1 above)
Comment (jeremyfiel) [reply curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md:17]: I don't necessarily agree with the hand holding once we're into "advanced" coursework. This should be the beginning of pushing a camper to figuring out how to progress with reading docs and such.  What do you think?

Comment (ojeytonwilliams) [reply curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md:17]: I think either we tell them what to do, or we point them to the docs so they can figure it out.  It's a little too much to expect people to realise that <em>render</em> refers to `res.render()`.

How about:

> Lastly, you should change your response to the request for the `/` route to use <code>res.render</code> to render the <code>index.pug</code> file from the <code>/views/pug/</code> directory.  The documentation for <code>res.render</code> can be found [here](https://expressjs.com/en/api.html#res.render) and the <code>view</code> path is relative to <code>/views</code>.
> 
>  Once you are successful your app will stop showing the message "Pug template is not defined." and will display "Looks like this page is being rendered from Pug into HTML!" instead.

It's a good point about "advanced" coursework, so if we want to introduce Glitch features (like 'Show') that should be done much earlier in the curriculum.

Comment (jeremyfiel) [reply curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md:17]: By the time they reach the advanced node section, they should have a good grasp on reading docs and exploring a bit. They've already gone through express setup with the basic node curriculum. Hopefully by then they've explored some documentation too. I'm not in favor of spoon feeding this late in the course.

The glitch feature is already introduced in the first express lesson on the basic node coursework. [Working Express Server](https://www.freecodecamp.org/learn/apis-and-microservices/basic-node-and-express/start-a-working-express-server) 
>Note: Be sure that your code works by looking at the logs, then see the results in your browser by clicking the ‘Show Live’ button if you are using Glitch.

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md:17]: As always, great discussion here, and @jeremyfiel brings up a good point about this being quite advanced in the curriculum. It's true that we don't want to spoon feed campers the correct answer and tell them exactly what to write and make it easy to copy and paste the answer in. But at the same time we want to keep them on the golden path as much as possible.

I'm in favor of @ojeytonwilliams's text above. It explains enough so campers won't need to go to the docs and look around but leaves that as in option for those who want to dive deeper into the `render` method.

Comment (ojeytonwilliams) [reply curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md:17]: This text https://github.com/freeCodeCamp/freeCodeCamp/pull/38572#discussion_r412317698?  If so, I'll edit my review suggestion.

Comment (jeremyfiel) [reply curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md:17]: @scissorsneedfoodtoo without giving them the actual method, do you think the tests are enough information not to spoon feed them this answer?  By this time, they've just completed the Chai testing curriculum prior to starting advanced node so they should be very familiar with gathering information from the test results. 
>text: Use the correct ExpressJS method to render the index page from the response.

If you are still leaning towards more descriptive text, please make a few grammatic changes: 
>you need to change your response to the request for the <code>/</code> route to render the file in <code>views/pug/~~index.pug~~</code>. <code>res.render()</code> takes a string of a file path **as an** argument. The path can be a relative path (relative to views), or an absolute path, and does not require **a file** extension. 

I bolded the changes.  I don't have access to modify the commit suggestion above.

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md:17]: @jeremyfiel, I went through the Chai testing curriculum and think that those descriptions are a bit too vague at times. Ideally each description would explain more about each method that's introduced. That's something we're trying to improve on in the new testing projects for the project based curriculum.

Sorry @ojeytonwilliams, yes, I meant the text from https://github.com/freeCodeCamp/freeCodeCamp/pull/38572#discussion_r412317698. I think it strikes a good balance of being descriptive without giving the answer away. But I also like elements of the descriptive text in https://github.com/freeCodeCamp/freeCodeCamp/pull/38572#discussion_r412950966.

Perhaps we can combine the two and get the best of both worlds. Something like:

```
Lastly, you should change your response to the request for the `/` route to use <code>res.render</code> to render the <code>index.pug</code> file in the <code>/views/pug/</code> directory.  `res.render()` takes a string of a file path as an argument. The path can be a relative path (relative to views), or an absolute path, and does not require a file extension.

Once you are successful your app will stop showing the message "Pug template is not defined." and will display "Looks like this page is being rendered from Pug into HTML!" instead.
```

There's enough information there that we probably don't need the link to the docs.

Comment (jeremyfiel) [reply curriculum/challenges/english/06-information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine.english.md]: it's all ready for you @scissorsneedfoodtoo !  thanks again!

**Comment by ojeytonwilliams** [comment]
@scissorsneedfoodtoo I'm merging this now, because @jeremyfiel pointed out that the instructions will be out of sync with the boilerplate.  Let me know if there's anything else needing changing and I'll create the PR.

**Review by ojeytonwilliams — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
@jeremyfiel nice work!  Thanks again for taking this on.

**Comment by WihanBarnard** [comment]
Hi all,

Thanks for the thorough discussion. I started working through the  advanced node and express section today and I could not pass the third test on the first challenge ("Use the correct ExpressJS method to render the index page from the response."). I see that there is one other camper who posted on the forum approx 3 days ago, and I see these changes were made quite recently.  The piece of code in the server.js from the glitch project is as follows:

```
app.route("/").get((req, res) => {
  //Change the response to render the Pug template
  res.render(process.cwd() + '/views/pug/index');
});
```

**Comment by jeremyfiel** [comment]
> Hi all,
> 
> Thanks for the thorough discussion. I started working through the advanced node and express section today and I could not pass the third test on the first challenge ("Use the correct ExpressJS method to render the index page from the response."). I see that there is one other camper who posted on the forum approx 3 days ago, and I see these changes were made quite recently (https://www.freecodecamp.org/forum/t/set-up-a-template-engine-keeps-failing/378920). The piece of code in the server.js from the glitch project is as follows:
> 
> ```
> app.route("/").get((req, res) => {
>   //Change the response to render the Pug template
>   res.render(process.cwd() + '/views/pug/index');
> });
> ```

you should be seeing this in the glitch project if you followed the link in the challenge "Setting up a template engine".  If you are using any other boilerplate then it's not updated as of yet. 
```js
app.route("/").get((req, res) => {
  //Change the response to render the Pug template
  res.send(`Pug template is not defined.`);
});
```
This was part of our discussion above where we mentioned avoiding node process methods in favor of relative path structuring.  While the node process is technically successful, we didn't include the ability to use it in the tests.
>The path can be a relative path (relative to views), or an absolute path, and does not require a file extension.

**Comment by ojeytonwilliams** [comment]
@jeremyfiel's pretty much covered it, but I can confirm that new projects created via the link in `As a reminder, this project is built upon the following starter project on Glitch` use the new boilerplate.  

@WihanBarnard if you've just started that section, I'd recommend starting a few project with the new boilerplate.

**Comment by WihanBarnard** [comment]
Thanks very much for the replies! I did continue the discussion on the forum, but though I could mention my finding here also:

When looking at the regex test for the third test:
```
assert.match(data, /res(\s+)?\.(r\w{5})\(('|")((\.{2})?)\/?(\w{3})\/(\w{5})('|")\)(;)?/gi, 'You successfully rendered the Pug template!');
```
It seems like the expected input is:
```
res.render('../pub/index');
```
When I use the following piece of code:
```
app.route("/").get((req, res) => {
  //Change the response to render the Pug template
  res.render('../pub/index');
});
```

I pass the test which is giving problems, but fail the final test stating Pug should be working. Also my page does not render.

**Comment by jeremyfiel** [comment]
> Thanks very much for the replies @jeremyfiel and @ojeytonwilliams ! I did continue the discussion on the forum, but though I could mention my finding here also:
> 
> When looking at the regex test for the third test:
> 
> ```
> assert.match(data, /res(\s+)?\.(r\w{5})\(('|")((\.{2})?)\/?(\w{3})\/(\w{5})('|")\)(;)?/gi, 'You successfully rendered the Pug template!');
> ```
> 
> UPDATE:
> This also satisfies the regex
> 
> ```
> res.render('pug/index');
> ```
> 
> (this was my first try, BUT I had to change
> 
> ```
> app.use("/public", express.static(process.cwd() + "/public"));
> ```
> 
> TO
> 
> ```
> 
> app.use("/public", express.static(process.cwd() + "/views"));
> ```
> 
> Is this intentional or is an update on the boilerplate needed?
copying from the forum thread...

The way the Express static method is outlined in this boilerplate is only for the CSS assets. If you change the static path to views, it will never display the css, which I supposed is meaningless for most of this curriculum because it’s focused on a node app (backend).
But I digress, the expectation is as you indicated in your update but the changes to the static resource are unnecessary.
the regex is expecting (.{2})? there “might be” 2 leading periods but it’s tricky because you’re right, it only works with using ../views/… or not at all. But the directions imply "relative to views".  Maybe we can take a look at modifying the test a bit.
Hope it helps!

**Comment by scissorsneedfoodtoo** [comment]
Thanks for bringing this up @WihanBarnard. Like @jeremyfiel and @ojeytonwilliams have mentioned, the boilerplate was updated recently, along with the regex test you pointed out above.

Maybe we should loosen up that test so it really is just checking for `res.render();` rather than trying to account for every possible path that can be passed as an argument. Perhaps something pretty permissive like `res(\s+)?\.(r\w{5})\(('|")?.*('|")?\)\;?`.

The last test for this challenge checks that the page is rendered properly whether by using Node processes or a relative path.

What do you all think?

**Comment by ojeytonwilliams** [comment]
Yep, the last test makes sure the right thing is rendered, so the main thing is to check is 'is render used?'.  How about a simplified version of @scissorsneedfoodtoo's:

```js
res\s*\.render\(.*?\)
```
i.e. specify render and allow basically everything else

or

```js
res\s*\.render\([^'"]*?('|").*(\1)\)
```

i.e. `res.render(possibly 'node.cwd() + ' followed by something in quotes)`.  This should match

```js
res.render('pug/index');
res.render(process.cwd() + '/views/pug/index');
```

**Comment by jeremyfiel** [comment]
i think we can go on forever with regex but let's relax the test to explicitly check only the method without revealing the name in the actual test.  anything after the method begins is left unchecked.

```js
res\s*\.r(\w{5})
```

**Comment by scissorsneedfoodtoo** [comment]
@ojeytonwilliams thank you for simplifying that. Either one is much more readable than the one I posted above.

@jeremyfiel, that could work, and I'm always up for simplifying wherever possible. My only concern with `res\s*\.r(\w{5})` is that something like `res.reeeee('test');` would still pass that test.

I think it would be better to test for the specific method we tell them to use in the instructions. Campers won't be shown the answer on the [challenge's page](https://www.freecodecamp.org/learn/information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine). If they don't use the `render` method, all they'll see is "Use the correct ExpressJS method to render the index page from the response."

**Comment by jeremyfiel** [comment]
> @ojeytonwilliams thank you for simplifying that. Either one is much more readable than the one I posted above.
> 
> @jeremyfiel, that could work, and I'm always up for simplifying wherever possible. My only concern with `res\s*\.r(\w{5})` is that something like `res.reeeee('test');` would still pass that test.
> 
> I think it would be better to test for the specific method we tell them to use in the instructions. Campers won't be shown the answer on the [challenge's page](https://www.freecodecamp.org/learn/information-security-and-quality-assurance/advanced-node-and-express/set-up-a-template-engine). If they don't use the `render` method, all they'll see is "Use the correct ExpressJS method to render the index page from the response."

well, i supposed they could but then the 4th test would fail and the page wouldn't load without using render because express doesn't have a reeeee method.  So they still wouldn't be able to pass the challenge.  I think the purpose of using the occurrence is to avoid them looking up the test answer explicitly.. even if they did, they still need to dig a bit more on the express docs to figure it out.. or by connecting the dots on the test message and the challenge instructions. if you prefer, i can change it.

**Pull request merged**